### PR TITLE
Make first rc

### PR DESCRIPTION
**Pull request by JelF**
Implemented, styled and documented Verifier (except flow documentation and README.md)

### CODE REVIEW

**Comment by akxcv** [comment]
I think you should configure Travis to only build the pull request once (e.x. only build pushes to "master" branch and PRs).

**Comment by JelF** [comment]
@akxcv travis made exactly one build when i pushed this branch

**Comment by akxcv** [comment]
@JelF and it made another one when you created this pull request.

**Comment by JelF** [comment]
@akxcv done

**Comment by JelF** [comment]
ill prefer to finalize work on this point and then add changes via separate PR's @akxcv, could you review entire repository in this PR?

File: .rubocop.yml [hunk 1]
```diff
@@ -14,3 +17,15 @@ Metrics/BlockLength:
 Lint/AmbiguousBlockAssociation:
   # REVIEW: why is that dangerous?
```
Comment (akxcv) [inline .rubocop.yml:18]: Not sure about the dangers, but I think that `some_method(a) { |val| puts val }` is more readable than `some_method a { |val| puts val }`.

File: lib/xverifier/applicator.rb [hunk 2]
```diff
@@ -2,77 +2,181 @@
 
 module XVerifier
   # @abstract implement `#call`
+  # Applicates objects called `applicable`
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "Applies "applicable" objects to given bindings" or smth like that

File: lib/xverifier/applicator.rb [hunk 3]
```diff
@@ -2,77 +2,181 @@
...
+  # (all objects are applicable, this typing made for usage purposes).
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "this naming(?) *is chosen* for usage(convenience?) purposes"

File: lib/xverifier/applicator.rb [hunk 4]
```diff
@@ -2,77 +2,181 @@
...
+  #
+  # This class uses ClassBuilder subsystem, start with .call method
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: Maybe just "system", not "subsystem"?
Also, "When reading the code, we suggest starting from '.call' method"

File: lib/xverifier/applicator.rb [hunk 5]
```diff
@@ -2,77 +2,181 @@
...
-    def self.build_class(applicable)
-      if applicable.is_a?(Symbol)
-        MethodExtractor
-      elsif applicable.is_a?(String)
-        InstanceEvaler
-      elsif applicable.respond_to?(:to_proc)
-        ProcApplicatior
-      else
-        Quoter
+    # Proxy is used when applicable itself is an Applicator.
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "when applicable is itself an instance of Applicator" sounds clearer

File: lib/xverifier/applicator.rb [hunk 6]
```diff
@@ -2,77 +2,181 @@
...
+    # It just delegates #call to it
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "It simply delegates "#call" method to the applicable."

File: lib/xverifier/applicator.rb [hunk 7]
```diff
@@ -2,77 +2,181 @@
...
+    # @example
+    #   Applicator.call(Applicator.build(:foo), binding_, context)
+    #   # => Applicator.call(:foo, binding_, context)
+    class Proxy < self
+      # @param applicable [Applicator]
+      # @return Proxy if applicable is Applicator
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "is an instance of Applicator"

File: lib/xverifier/applicator.rb [hunk 8]
```diff
@@ -2,77 +2,181 @@
...
       end
-    end
-
-    def self.call(applicable, binding, context)
-      build_class(applicable).new(applicable).call(binding, context)
-    end
 
-    def initialize(applicable)
-      self.applicable = applicable
+      # @param binding_ [#instance_exec] target to applicate applicable
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "target to apply the applicable to"
It's beginning to sound weird, but that's just a consequence of naming.

File: lib/xverifier/applicator.rb [hunk 9]
```diff
@@ -2,77 +2,181 @@
...
+      # @param context additional info to send it to applicable
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "to send to the applicable", you don't need the "it" here

File: lib/xverifier/applicator.rb [hunk 10]
```diff
@@ -2,77 +2,181 @@
...
+      end
     end
 
-    # @!method call(binding, context)
-    #   @abstract
-    #   @param binding [#instance_exec]
-    #   @param context
-
-    private
+    # MethodExtractor is used when applicable is a symbol.
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "the applicable"
Also, "Symbol"

File: lib/xverifier/applicator.rb [hunk 11]
```diff
@@ -2,77 +2,181 @@
...
+    # It extracts extracts method from binding_ and executes it on binding_
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "It extracts a method from binding_ and then executes it in its context" or smth

File: lib/xverifier/applicator.rb [hunk 12]
```diff
@@ -2,77 +2,181 @@
...
+      def self.build_class(applicable)
+        self if applicable.is_a?(Symbol)
+      end
 
-    def invoke_lambda(lambda, binding, context)
-      if lambda.arity.zero?
-        binding.instance_exec(&lambda)
-      else
-        binding.instance_exec(context, &lambda)
+      # @param binding_ [#instance_exec] target to applicate applicable
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: Same as earlier

File: lib/xverifier/applicator.rb [hunk 13]
```diff
@@ -2,77 +2,181 @@
...
+          binding_ = binding_.dup
+          binding_.local_variable_set(:context, context)
+          binding_.eval(applicable, *caller_line)
         else
-          binding.instance_eval(applicable, *caller_line)
+          binding_.instance_eval(applicable, *caller_line)
         end
       end
 
+      # @return [String, Integer] file and line where `Applicator.call` called
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "was called", or "is called", or "is being called"... Depending on what you actually meant

File: lib/xverifier/applicator.rb [hunk 14]
```diff
@@ -2,77 +2,181 @@
...
+
+      # @param binding_ [#instance_exec] target to applicate applicable
+      # @param context additional info to send it to applicable
+      # @return application result
+      def call(binding_, context)
+        invoke_lambda(applicable.to_proc, binding_, context)
       end
     end
 
+    # Quoter is used when there is no other way to applicate applicatable.
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: Here and everywhere else, change "applicate" to "apply".

File: lib/xverifier/applicator.rb [hunk 15]
```diff
@@ -2,77 +2,181 @@
...
     end
+
+    extend ClassBuilder::Mixin
+
+    self.buildable_classes =
+      [Proxy, MethodExtractor, InstanceEvaluator, ProcApplicatior, Quoter]
+
+    attr_accessor :applicable
+
+    # Applicates applicable on binding_ with context
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: And "applicates" to "applies".

File: lib/xverifier/applicator.rb [hunk 16]
```diff
@@ -2,77 +2,181 @@
...
+    # @todo add @see #initialize when it's todo done
+    # @param applicable [applicable]
+    #   see examples in sublcasses defenitions
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "in subclass definitions"
Or, "in definitions of subclasses"

File: lib/xverifier/applicator.rb [hunk 17]
```diff
@@ -2,77 +2,181 @@
...
+    # @param binding_ [#instance_exec]
+    #   where should applicable be applied. It could be either a generic object,
+    #   where it would be `instance_exec`uted, or a binding_
+    # @param context
+    #   geneneric data you want to pass to applicable function.
+    #   If applicable would not accept params, it would not be sent
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "If applicable cannot accept params, the data will not be sent"

File: lib/xverifier/applicator.rb [hunk 18]
```diff
@@ -2,77 +2,181 @@
...
+    # @param [Applicator] other
+    # @return [Boolean] true if applicable matches, false otherwise
+    def ==(other)
+      applicable == other.applicable
+    end
+
+    # @!method call(binding_, context)
+    #   @abstract
+    #   Applicates applicable on binding_ with context
+    #   @param binding_ [#instance_exec] biding would be used in application
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "binding to be used for application"

File: lib/xverifier/applicator.rb [hunk 19]
```diff
@@ -2,77 +2,181 @@
...
+    #   @param context param would be passed if requested
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "param that will be passed if requested"

File: lib/xverifier/applicator.rb [hunk 20]
```diff
@@ -2,77 +2,181 @@
...
+    #   @return application result
+
+    private
+
+    # invokes lambda basing on it's arity
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "based on its arity"

File: lib/xverifier/applicator.rb [hunk 21]
```diff
@@ -2,77 +2,181 @@
...
+    # @param [Proc] lambda
+    # @param binding_ [#instance_exec] binding_ would be used in application
+    # @param context param would be passed if lambda arity > 0
+    # @return invokation result
```
Comment (akxcv) [inline lib/xverifier/applicator.rb]: "invocation"

File: lib/xverifier/applicator_with_options.rb [hunk 22]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module XVerifier
+  # An applicator with usefull options
```
Comment (akxcv) [inline lib/xverifier/applicator_with_options.rb]: useful

File: lib/xverifier/applicator_with_options.rb [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+  # @example if
+  #   ApplicatorWithOptions.new(action, if: true)
+  # @example unless
+  #   ApplicatorWithOptions.new(action, unless: false)
+  # @attr action [Applicator]
+  #   main action to applicate on call
+  # @attr if_condition [Applicator]
+  #   main action only applicate if this applicates to truthy value
```
Comment (akxcv) [inline lib/xverifier/applicator_with_options.rb]: "...only applies if this evaluates..."

File: lib/xverifier/applicator_with_options.rb [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+    end
+
+    # Applicates main action if if_condition applicated to truthy value
+    # and unless_condition applicated to falsey value
+    # @param binding_ [#instance_exec]
+    #   binding to applicate (see Applicator)
+    # @param context
+    #   generic context to applicate (see Applicator)
+    # @return main action application result
+    # @return [nil] if conditions check failed
```
Comment (akxcv) [inline lib/xverifier/applicator_with_options.rb]: condition checks

File: lib/xverifier/class_builder.rb [hunk 25]
```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+module XVerifier
+  # ClassBuilder is something like Uber::Builder, but it uses
+  # inverted dependencies. It means, that child classes are
```
Comment (akxcv) [inline lib/xverifier/class_builder.rb]: dependency injection / dependency inversion?

File: lib/xverifier/class_builder.rb [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+  # deciding to be or not to be used
```
Comment (akxcv) [inline lib/xverifier/class_builder.rb]: "It means that child classes decide whether they should be built or not"

File: lib/xverifier/class_builder.rb [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+  # @attr klasses [Array(Class)]
+  #   klasses to iterate during search of most suitable
```
Comment (akxcv) [inline lib/xverifier/class_builder.rb]: classes

File: lib/xverifier/class_builder.rb [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+  class ClassBuilder
+    # Mixin provides usefull methods to integrate builder subsystem.
+    # Feel free to override or just never include it.
+    # @attr_writer [Array(Class)] buildable_classes
+    #   Array of classes which would be checked if they
```
Comment (akxcv) [inline lib/xverifier/class_builder.rb]: will

File: lib/xverifier/class_builder.rb [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+    #   suite constructor arguments. Order maters
```
Comment (akxcv) [inline lib/xverifier/class_builder.rb]: matters

File: lib/xverifier/class_builder.rb [hunk 30]
```diff
@@ -0,0 +1,55 @@
...
+    module Mixin
+      # Array of classes which would be checked if they
+      # suite constructor arguments. Order maters
+      # @param klasses [Array(Class)]
+      def buildable_classes=(klasses)
+        @class_builder = ClassBuilder.new(klasses).freeze
+      end
+
+      # Default implementation if build_class. Feel free to change
```
Comment (akxcv) [inline lib/xverifier/class_builder.rb]: of

File: lib/xverifier/verifier.rb [hunk 31]
```diff
@@ -10,12 +10,17 @@ module XVerifier
   # @attr messages [Array]
   #   Array to collect all messages yielded by verifier
```
Comment (akxcv) [inline lib/xverifier/verifier.rb]: Array with all messages yielded by the verifier

File: lib/xverifier/verifier/applicator_with_options_builder.rb [hunk 32]
```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module XVerifier
+  class Verifier
+    # Builds ApplicatorWithOptions from different invokation styles.
```
Comment (akxcv) [inline lib/xverifier/verifier/applicator_with_options_builder.rb]: invocation

File: lib/xverifier/verifier/applicator_with_options_builder.rb [hunk 33]
```diff
@@ -0,0 +1,74 @@
...
+    # @api private
+    # @attr base [Class]
+    #   class for which applicator_with_options should be built
+    # @attr args [Array]
+    #   array of arguments invoked with
```
Comment (akxcv) [inline lib/xverifier/verifier/applicator_with_options_builder.rb]: an array of arguments to invoke methods with ?

File: lib/xverifier/verifier/applicator_with_options_builder.rb [hunk 34]
```diff
@@ -0,0 +1,74 @@
...
+    # @attr block [Proc]
+    #   block invoked with
```
Comment (akxcv) [inline lib/xverifier/verifier/applicator_with_options_builder.rb]: -||-

File: lib/xverifier/verifier/applicator_with_options_builder.rb [hunk 35]
```diff
@@ -0,0 +1,74 @@
...
+      # @!method self.call(base)
+      # transforms `verify` arguments to class attributes
+      # and invokes calculation
+      # @raise [ArgumentError]
+      # @return [ApplicatorWithOptions] resulting applicator_with_options
+      def self.call(base, *args, &block)
+        new(base, args, block).call
+      end
+
+      # Tries different invokation styles untill one matches
```
Comment (akxcv) [inline lib/xverifier/verifier/applicator_with_options_builder.rb]: invocation; until

File: lib/xverifier/verifier/applicator_with_options_builder.rb [hunk 36]
```diff
@@ -0,0 +1,74 @@
...
+      #   verify SubVerifier
+      # @example incorrect
+      #   verify Class
+      # @raise [ArgumentError]
+      # @return [ApplicatorWithOptions]
+      def try_nesting
+        verifier, *rest = args
+        return unless verifier.is_a?(Class)
+        raise ArgumentError, <<~ERROR unless verifier < base
+          Nested verifiers should inherit verifier they nested in
```
Comment (akxcv) [inline lib/xverifier/verifier/applicator_with_options_builder.rb]: "...should inherit from the verifier they are nested in"

File: spec/xverifier/applicator_spec.rb [hunk 37]
```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+describe XVerifier::Applicator do
+  subject(:applicator) { described_class.build(applicable) }
+
+  let(:binding_class) do
+    Class.new { define_method(:foo) { |*| } }
```
Comment (akxcv) [inline spec/xverifier/applicator_spec.rb]: What on Earth is this magic?

File: spec/xverifier/class_builder_spec.rb [hunk 38]
```diff
@@ -0,0 +1,25 @@
...
+
+describe XVerifier::ClassBuilder do
+  let(:klass) do
+    # This extend is used in all specs for class_double purposes.
+    # You are not forced to extend mixin if you implement `.build_class`
+    # in all `klasses`
+    Class.new { extend XVerifier::ClassBuilder::Mixin }
+  end
+
+  let(:unbuildabe) { class_double(klass, build_class: nil) }
```
Comment (akxcv) [inline spec/xverifier/class_builder_spec.rb]: unbuildable

File: spec/xverifier_spec.rb [hunk 39]
```diff
@@ -0,0 +1,14 @@
...
+    it 'is well-formated' do
+      expect(XVerifier::VERSION).to match(/\A0\.\d+\.\d+\.\d+(_.+)?\z/)
+    end
+
+    it 'is released' do
+      pending 'WIP'
+      expect(XVerifier::VERSION).to match(/\A\d+\.\d+\.\d+(_.+)?\z/)
+    end
+  end
+end
```
Comment (akxcv) [inline spec/xverifier_spec.rb:14]: ❤️

**Review by akxcv — CHANGES_REQUESTED** [review]
Overall looks good! 👍

File: .rubocop.yml (hunk 1 above)
Comment (JelF) [reply .rubocop.yml:18]: It is about following syntax:
```
spec/xverifier/verifier_spec.rb:55:9: W: Parenthesize the param lambda to make sure that the block will be associated with the verify method call.
        verifier.verify ->(context) { message!(context) }
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: lib/xverifier/applicator.rb (hunk 13 above)
Comment (JelF) [reply lib/xverifier/applicator.rb]: i also forgot to make a spec for this

File: lib/xverifier/verifier/applicator_with_options_builder.rb (hunk 34 above)
Comment (JelF) [reply lib/xverifier/verifier/applicator_with_options_builder.rb]: ???

Comment (akxcv) [reply lib/xverifier/verifier/applicator_with_options_builder.rb]: I meant "same as above".
"a block to invoke methods with"

File: .rubocop.yml (hunk 1 above)
Comment (akxcv) [reply .rubocop.yml:18]: It wants you to put parens around the lambda...
Yeah, I guess we can do without this rule.

File: lib/xverifier/applicator.rb (hunk 20 above)
Comment (JelF) [reply lib/xverifier/applicator.rb]: Doesn't it mean like "select some lambda"
Used: "invokes lambda respecting it's arity"

File: lib/xverifier/applicator_with_options.rb (hunk 23 above)
Comment (JelF) [reply lib/xverifier/applicator_with_options.rb]: evaluates is more generic and does not describe, that this stuff would be applicated using the rules of application

File: lib/xverifier/verifier/applicator_with_options_builder.rb (hunk 33 above)
Comment (JelF) [reply lib/xverifier/verifier/applicator_with_options_builder.rb]: it changes sence, but should be okish

Comment (JelF) [reply lib/xverifier/verifier/applicator_with_options_builder.rb]: well, not in this place. I'll try to reword:
array of arguments `Verifier.verify` invoked with

**Comment by JelF** [comment]
@akxcv can you re-check it?

File: .rubocop.yml [hunk 40]
```diff
@@ -14,3 +17,15 @@ Metrics/BlockLength:
...
   Enabled: false
+
+RSpec/FilePath:
+  CustomTransform:
+    XVerifier: xverifier
+
+RSpec/MessageSpies:
+  # Spies are not always usable, i.e. we wanna to stub something we haven't create
```
Comment (akxcv) [inline .rubocop.yml]: 1) "wanna" = "want to"
2) "... haven't created"

**Review by akxcv — COMMENTED** [review]
Alright, let's correct this one line and then I'll make a pull request with grammar corrections for the README (eventually, if I find any mistakes).

**Review by akxcv — APPROVED** [review]

**Pull request merged**